### ISSUE TITLE
Why result depends on column order?

### ISSUE THREAD

**Issue by mrgloom** [catboost/catboost#334]
catboost version: {0.8.1}
Operating System: {Ubuntu 14}
CPU: {CPU}

I'm trying to solve some classification problem, code like:

```
model = CatBoostClassifier(loss_function='Logloss',
                           thread_count=multiprocessing.cpu_count(),
                           random_seed = 2018,
                           #depth = 2,
                           logging_level='Verbose')

model.fit(train_x, train_y)
pred_y = model.predict(test_x)
```
And as I can see order of columns in `train_x` pandas dataframe affects results of classification, so my question is why result depends on column order?

**Comment by m-rahimi** [comment]
Catboost selects some columns randomly for each iteration and changing the order of columns affects the selection of the Catboost.  in Xgboost if you set two parameters colsample_bylevel and colsample_bytree to 1 changing the order doesn't affect the result. I am sure you can find similar parameters in Catboost.

**Comment by annaveronika** [comment]
This is because of randomness. There are two reasons for this behavoir, both of them are expected.
First one is rsm or colsample_bylevel. This is random selection of features for next split.
The second one is random_strength - this is a random noise that is added to scores to reduce overfitting.
If you don't want any randomness you can set rsm == 1 and random_strength == 0, this should lead to independence of column order. But we are planning to add some other things that will lead to difference if you have different feature order. So please rely on this behavoir.

**Comment by Serebryanskiy** [comment]
Does this mean that we shouldn't really use optuna for Catboost, since even if we find optimal parameters, we couldn't rely on retraining the model with these params, since due to randomness the results will be different every time?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
please specify the following options on cpu to make training deterministic
`has_time=true, seed=0, bootstrap_type='No', random_strength=0`

**Comment by tle4336** [comment]
@Evgueni-Petrov-aka-espetrov I tried your suggestions and still got different prediction results on the same testing dataset using the same training dataset, and I trained the model on CPU. Any idea on how to really fix this problem?
